binderhub/events.py

```diff
@@ -0,0 +1,97 @@
+"""
+Emit structured, discrete events when various actions happen.
+"""
+from traitlets.config import Configurable
+
+import logging
+from datetime import datetime
+from pythonjsonlogger import jsonlogger
+from traitlets import TraitType
+import json
+import six
+
+
+class Callable(TraitType):
+    """
+    A trait which is callable.
+
+    Classes are callable, as are instances
+    with a __call__() method.
+    """
+    info_text = 'a callable'
+    def validate(self, obj, value):
+        if six.callable(value):
+            return value
+        else:
+            self.error(obj, value)
+
+
+class EventLog(Configurable):
+    """
+    Send structured events to a logging sink
+    """
+    handlers_maker = Callable(
+        None,
+        config=True,
+        allow_none=True,
+        help="""
+        Callable that returns a list of logging.Handler instances to send events to.
+
+        When set to None (the default), events are discarded.
+        """
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        def _skip_message(record, **kwargs):
```

Comment:
This can be a top-level instance or staticmethod. That way the definition doesn't need to be evaluated on instantiation of the object.